lib/Analysis/OptimizeRelinearizationAnalysis/OptimizeRelinearizationAnalysis.cpp

```diff
@@ -0,0 +1,371 @@
+#include "lib/Analysis/OptimizeRelinearizationAnalysis/OptimizeRelinearizationAnalysis.h"
+
+#include "lib/Dialect/BGV/IR/BGVDialect.h"
+#include "lib/Dialect/BGV/IR/BGVOps.h"
+#include "llvm/include/llvm/ADT/DenseMap.h"             // from @llvm-project
+#include "llvm/include/llvm/ADT/TypeSwitch.h"           // from @llvm-project
+#include "llvm/include/llvm/Support/Debug.h"            // from @llvm-project
+#include "mlir/include/mlir/Dialect/Func/IR/FuncOps.h"  // from @llvm-project
+#include "mlir/include/mlir/IR/Operation.h"             // from @llvm-project
+#include "mlir/include/mlir/IR/Value.h"                 // from @llvm-project
+#include "ortools/linear_solver/linear_solver.h"  // from @com_google_ortools
+
+// The key basis degree describes the highest degree in the key basis, i.e., a
+// key basis of (1, s, s^2) has degree 2, while (1, s, s^2, s^3) has degree 3.
+//
+// For now, fix the key basis degree bound to 3. Could make it a pass flag
+// later, or introduce costs for high-degree operations and let it be
+// unconstrained.
+constexpr int MAX_KEY_BASIS_DEGREE = 3;
+constexpr int INITIAL_KEY_BASIS_DEGREE = 1;
+constexpr int IF_THEN_AUX = 100;
+
+using namespace operations_research;
+
+namespace mlir {
+namespace heir {
+
+#define DEBUG_TYPE "OptimizeRelinearizationAnalysis"
+
+std::string nameAndLoc(Operation *op) {
+  std::string varName;
+  llvm::raw_string_ostream ss(varName);
+  ss << op->getName() << "_" << op->getLoc();
+  return ss.str();
+}
+
+bool hasCiphertextType(ValueRange range) {
+  return std::any_of(range.begin(), range.end(), [](Value value) {
```

Comment:
```suggestion
  return llvm::any_of(range, [](Value value) {
```

here and below in the transform (unless there was a reason to avoid the LLVM helpers?)

Reply:
Fixed in 140fdbdd